### ISSUE TITLE
[Feature request] option to automatically install a package when using `using`

### ISSUE THREAD

**Issue by gitboy16** [JuliaLang/Pkg.jl#3027]
As suggested by @KristofferC in the following link:
https://discourse.julialang.org/t/using-package-install-automatically/70896/5
I am opening an issue so that maintainers can discuss whether they deem it useful or not to have an option to automatically install a package when using `using`. That option would bypass the prompt message asking whether the user wants to install the package or not.

Just speaking for myself here (i.e. not for other users), but when I use `using X`, I will probably answer yes  99% of the time to the question "do you want to install package X".

That option could be implemented via macro or env variable, I don't really have a strong view on it.

I just think it is a nice to have option.

Thank you

**Comment by IanButterworth** [comment]
In 1.8 you get the options y/n/o where `o` lets you choose the environment you want to install into. If you're not in the main environment this can be useful for installing packages like Revise or BenchmarkTools into the main env that you don't want in your current project, but will want available. 

There could be an env var to automatically select `y` but it would side-step that functionality, which seems a shame. But given it would be opt-in, then it sounds reasonable, as long as new users aren't persuaded to add the env var by default.

**Comment by KristofferC** [comment]
In my original test implementation of installing during `using` I had that pressing `Y` would remember your choice (while `y` would not). But you would still have to make one active choice. Perhaps an env variable so that you can set it even before the first `using` is a better idea.

**Comment by IanButterworth** [comment]
One thought.. would removing the need to confirm open users up more to typo issues, malicious or not. I guess they'd still see the package installation, so that would give them pause?

**Comment by KristofferC** [comment]
Well, that's something that is up to the user. Just by installing a package (or doing an update), you are already opening yourself up to some level of possible maliciousness.

**Comment by IanButterworth** [comment]
Yeah I guess I just meant that if there's no confirm, they might not be aware they're installing a new package. But the messaging probably makes it clear.

**Comment by gitboy16** [comment]
Would you be ok with a macro like the following?

```julia
@add_pkg using DataFrames, CSV
```
We can call the macro as you wish and it would need to be available in Julia when the session starts.

The same macro coukd install from a local directory.

```julia
@add_pkg "C:/temp/MyPackage" using MyPackage
```
Similarly we can have `@dev_pkg` macro.

I am not fan of environment variable. I always forget about them or to set them up. I prefer to have everything in the code.

Alternatively we could have some like:
```julia
using install=true DataFrames, CSV
```

Is that something you would consider?

**Comment by gitboy16** [comment]
Or we can simply have a macro that install packages at the beginning of the script without the need to load Pkg:
```julia
@add_pkg DataFrames, CSV
using DataFrames
import CSV
```
The macro would check if the packages are install or not and if not then install them automatically without prompt message requesting input from the user every time a package needs to be installed.

**Comment by GHTaarn** [comment]
I tried to think who would actually use a feature like this, and came up with 3 scenarios:

1. An experienced Julia user starts a simulation on a Friday afternoon, leaves for the weekend and comes back Monday morning to a prompt asking him if he wants to Install the `DataFrames` package, instead of having a set of simulation results.
2. A teacher has a 2 hour hands on class with some students that have little programming experience and know nothing about Julia. In order to simplify their experience she would like to avoid the complication of explaining packages.
3. An employee in the IT-department is making a program for the sales department. He/she does not want the program to have the complication of answering questions about installing packages. This can of course be done without this new feature, but this feature could greatly reduce the number of lines of code needed to do this, which would hopefully both save time and reduce the probability of bugs.

Scenarios 1 and 2 would benefit from something simple that can be done either just before Julia starts or just after. @KristofferC s suggestion of an env variable would be just perfect for this, you can put it in your startup.jl , you can set it in your .bashrc , you can use it with the -i -e commandline options, you can type it manually in the REPL and a lot more. To use it in the example given by @gitboy16 you would write something like:

```julia
ENV["JULIA_PKG_USING_AUTOINSTALL"] = "yes"
using DataFrames
import CSV
```

Like @IanButterworth , I also get a gut reaction that creating this feature could have some bad consequences. Maybe Julia could get the blame for a bad user experience caused by this. Would it be a good idea to explicitly warn about the dangers of using the feature in the documentation of it?

I also want to say that I am not sure that I would use this feature personally, I like the current behaviour, but I do see situations where others could find it useful.

**Comment by gitboy16** [comment]
The only reason why I prefer a macro over an environment variable is because it forces the user to add it to the script/code and makes it clear to whoever is reading the code. Anyway feel free to close if you don't care about it.

**Comment by NilsWildt** [comment]
Something I've been using quite a while is https://github.com/mossr/AddPackage.jl
However, some command to pull just whatever I need from registry to run my script would be convenient.

**Comment by oschulz** [comment]
I've been using this now in Colab notebooks ([example link](https://colab.research.google.com/drive/1351EjrlNOEbBtBsEqlGfXRiNKamEbDCE)):

```julia
macro installpkgs(expr)
    function extract_pkgnames(ex)
        if !(ex isa Expr)
            Symbol[]
        elseif ex.head in (:using, :import)
            [(p.head == :. ? only(p.args) : only(p.args[1].args))::Symbol for p in ex.args]
        elseif ex.head in (:block, :let)
            vcat(extract_pkgnames.(ex.args)...)
        elseif ex.head in (:if, :elseif, :for, :while)
            vcat(extract_pkgnames.(ex.args[2:end])...)
        else
            Symbol[]
        end
    end
    pkgnames = extract_pkgnames(expr)
    :(import Pkg; Pkg.add($(String.(pkgnames)); preserve=Pkg.PRESERVE_TIERED_INSTALLED); $expr)
end
```

So one can just do

```julia
@installpkgs begin
    using CairoMakie
    import JSON
end
```

It's basically a more powerful version of `AddPackage.@add`, it supports constructions like (just a test example, not very sensible)

```julia
@installpkgs begin
    using JSON, StructArrays
    begin
        if 1 == 2
            import Tables
        elseif 1 == 3
            import ChangesOfVariables as CoV, ArraysOfArrays as AoA
        else
            let
                using Adapt: adapt, adapt_structure
            end
        end
    end
end
```

If we would add a macro like this (I expect we can find better and more stable code to extract package names from expressions in Pluto) to Julia Base, we'd have a convenient way to make standalone Julia scripts and Jupyter/Colab notebooks, especially for didactic settings where "zero setup" is key.

**Comment by oschulz** [comment]
I've registered a little package AutoPkg now that provides `@autopkg` (an improved version of the `@installpkgs` macro above), to make it easy to add and use packages in one go in Colab notebooks ([example link](https://colab.research.google.com/drive/1351EjrlNOEbBtBsEqlGfXRiNKamEbDCE)) and the like.

Just have a cell like

```julia
import Pkg; Pkg.add("AutoPkg"); using AutoPkg
```

at the top of the notebook, then you can just do

```julia
@autopkg begin
    using Distributions
    using Makie
    import CairoMakie
end
```

but also complex stuff like

```julia
@autopkg begin
    using SomePackage, SomeOtherPackage
    import YetAnotherPackage

    if some_condition
        using OneMorePackage: some_feature
    end
end
```

On Google Colab, `@autopkg` installs packages directly (since the notebooks run in isolated containers anyway), otherwise it activates a temporary Julia project env. Also useful to quickly try out packages on the REPL and so on.
